### PR TITLE
msgraphforsharepoint:Feature - Add support for subsites and adding items to lists

### PR DESCRIPTION
**Pull request by import-au**
Integration tests pipeline: https://cd.splunkdev.com/phantom-apps/app-tests/-/pipelines/6473562

Static tests pipeline will pass once https://cd.splunkdev.com/phantom/app_release/-/merge_requests/1996 is merged

Please ensure your pull request (PR) adheres to the following guidelines:

- Please refer to our contributing documentation for any questions on submitting a pull request, link: [Contribution Guide](https://github.com/splunk-soar-connectors/.github/blob/main/.github/CONTRIBUTING.md)

#### Please check if your PR fulfills the following requirements:
- [x] Testing of all the changes has been performed (for bug fixes / features)
- [x] The readme.html has been reviewed and added / updated if needed (for bug fixes / features)
- [x] Use the following format for the PR description: `<App Name>: <PR Type> - <PR Description>`
- [x] Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.
- [x] Verify all checks are passing.
- [x] Do NOT use the `next` branch of the forked repo. Create separate feature branch for raising the PR.
- [x] Do NOT submit updates to dependencies unless it fixes an issue.
#### Please check the type of change your PR introduces:
- [ ] New App
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation
- [ ] Other (please describe): 

## Security Considerations (REQUIRED)
- [ ] If you are exposing any endpoints using a [REST handler](https://docs.splunk.com/Documentation/SOAR/current/DevelopApps/RESTHandlers), 
  please document them in the `readme.html`.
- [ ] If this is a new connector or you are adding new actions
    - [ ] Please document in the `readme.html` all methods (eg, OAuth) used to authenticate 
      with the service that the connector is integrating with.
    - [ ] If any actions are unable to run on SOAR Cloud, please document this in the `readme.html`.
- [ ] Are you introducing any new cryptography modules? If yes, please elaborate their purpose:
- [ ] Are you are accessing the file system? If yes, please verify that you are only accessing paths returned through 
the [Vault](https://docs.splunk.com/Documentation/SOARonprem/5.2.1/DevelopApps/AppDevAPIRef#Vault) API.
- [ ] Are you are marking code to be ignored by Semgrep with [`nosemgrep`](https://semgrep.dev/docs/ignoring-files-folders-code/#ignoring-code-through-nosemgrep)?
If yes, please provide justification in an additional comment next to the ignored code.

  

## Release Notes (REQUIRED)
- Provide release notes as part of the PR submission which describe high level points about the changes for the upcoming GA release.
- Add optional support for sub-sites in Sharepoints know as Teams or Groups
- Add support for adding items to lists

## What is the current behavior? (OPTIONAL)
- Describe the current behavior that you are modifying.
- Sharepoint app only supports lists that exist at the root of the site.

## What is the new behavior? (OPTIONAL)
- Describe the behavior or changes that are being added by this PR.
- Adding the ability to modify sites that exist underneath the site. Commonly known a groups or teams.

## Other information (OPTIONAL)
- Any other information that is important to this PR such as screenshots of how the component looks before and after the change.
- The output for the action I added needs to be added to the README.md, wasn't sure how that was generated.

## Pay close attention to (OPTIONAL)
- Any specific code change or test case points which must be addressed/reviewed at the time of GA release.
- I validated with our tenant that both the root site and subsites work, not sure about other tenants as I can't test outside my tenant.

## Screenshots (if relevant)

---
Thanks for contributing!

### CODE REVIEW

**Comment by import-au** [comment]
MS Documentation for API endpoints added:
https://docs.microsoft.com/en-us/graph/api/listitem-update?view=graph-rest-1.0&tabs=http
https://docs.microsoft.com/en-us/graph/api/listitem-create?view=graph-rest-1.0&tabs=http

**Comment by pzhou-splunk** [comment]
Thanks for your submission! We have a total of 7 PRs open right now, and we’re working hard on all of them! We'll take a look as soon as we can.

**Comment by achandya-crest** [comment]
Hello @import-au,
We observed the parameter **item** of **add_item** and **update_item** actions that accepts JSON string for creating or updating item properties. We tried updating a few values and observed that the value in the parameter is expected in the fields key as mentioned below: 
{
    "fields":{
        "Title": "test",
        "Color": "Red"
    }
}
and on code side this input will be passed in data of API request, only then it will be updated in item. It is nowhere mentioned that the values to be added or updated in item should be passed in fields key in dictionary. 


Do you want us to hardcode **fields** key in action code so that the users can directly pass this below input in item parameter: 
{
    "Title": "test",
    "Color": "Red"
}
We will then pass this input to fields dictionary from code and then make the API call.

Could you please provide your inputs on this?

**Comment by achandya-crest** [comment]
We have observed that multiple fields can be added or updated in an item according to the template of the item. **Could you please tell us which all fields are you expecting should be added or updated?**
Coz we see a few fields like add attachments, add images from app UI, so we were wondering if they are expected to be modified or not?

**Comment by devals-crest** [comment]
Hello @import-au 
We tried using the SharePoint Group ID parameter and provided a valid Group ID as an input but it did not make any difference in the actions for which it is used. Could you please suggest what is the significance of this parameter so that we can know how it is to be tested?
Thanks in advance!

**Comment by import-au** [comment]
Hey all,

@achandya-crest The JSON fields are purely based on the fields in the list. If there is a "Description" column in the list, there will be a "Description" key in the JSON. For this reason, I did not hardcode any values. My inputs look something like this:

`{"fields":{"Title":"This is a title", "Details":"This is the details", "Region":"North America", "Event Type":"This is a type"}}`
Each of these fields under the nested dict are dependent on the sheet. There are other dicts that can be added here other than `fields` as well.


@devals-crest The group ID are like sub teams within a larger Sharepoint site ID. The "Sharepoint Site ID" is a parent organization, like the company I work for, the SharePoint Group ID is a subsite or a team within the Site ID, like the security organization. It's used in the API to delineate where your lists are for updating.

**Comment by devals-crest** [comment]
Hello @import-au 
We are planning to avoid passing “fields” from the code side as suggested by you. Also, the properties that could be added or updated by the JSON format input, will get reflected on the UI.

As you have added the SharePoint Group ID parameter in the asset configuration parameters, shall we validate it in the test connectivity as well? 
It is mentioned in the API doc: 
Screenshot 2022-08-05 at 7 20 30 PM

So we are planning to add the below snippet in the test connectivity code:

```
if self._group_id:
            endpoint = "/groups/{}/sites/root".format(self._group_id)
        else:
            endpoint = /sites/root
```

Could you please provide your input on it?

-> Could you please explain to us the significance of the SharePoint Group ID parameter so that we can test it on our environment? We tried with various valid groups ids but got the same results of list lists, list sites, etc actions as we got when we ran the actions without the group id parameter. 
It would be great if you provide us with some steps/details/screenshots on how to work with this parameter?
Thanks in advance!

**Comment by import-au** [comment]
Yeah, I have something similar implemented in the connector [here](https://github.com/splunk-soar-connectors/msgraphforsharepoint/pull/3/files#diff-01848deb66b70cb7ff587476d06d70bf2a9b383368c5f1217b19286789a51752R798-R801) so it would be a good idea to do the same in test connectivity.

For the group ID, you would need to create a subsite in your organization, and then a list inside that subsite. I'd rather not share my requests here due to the public nature of a PR, but I am happy to show it in a zoom session or something a bit less public?

**Comment by achandya-crest** [comment]
Hello @import-au, Thanks for joining on the call and helping us conclude quickly.

As discussed in the call, we will **not add the Sharepoint Group ID parameter to the asset configuration parameters**. 

The major reason for this is in the latest release, the **Sharepoint Site ID** parameter was an **optional** parameter, which could be fetched from the **list_sites** action.
This is also mentioned in the documentation of the app. Also, the reason for not adding the Sharepoint Group ID parameter is, that it was added with the intention to add the support of sub-sites, but the **Sharepoint Site ID parameter already satisfies the same requirement**.

Attaching the screenshot of the app documentation where this behavior is mentioned:
Screenshot 2022-08-09 at 8 27 07 PM

File: msgraphforsharepoint.json [hunk 1]
```diff
@@ -1095,6 +1317,725 @@
...
+        },
+        {
+            "action": "update item",
+            "identifier": "update_item",
+            "description": "Update an item in a list on a SharePoint Site",
+            "type": "generic",
+            "read_only": false,
+            "parameters": {
+                "list": {
+                    "description": "Title or ID of the list to update it's item",
```
Comment (splunk-jessica) [inline msgraphforsharepoint.json:1683]:
```suggestion
                    "description": "Title or ID of the list to which the item belongs",
```

**Review by splunk-jessica — APPROVED** [review]

**Review by ishans-crest — APPROVED** [review]

**Pull request merged**